### ISSUE TITLE
kfctl semantics replace init & generate with build  and also add apply

### ISSUE THREAD

**Issue by jlewi** [kubeflow/kubeflow#3518]
Should we change the semantics of kfctl and replace the init and generate commands with

```
kfctl build --config=path to config file
```

and 

```
kfctl build
```
   * Look for app.yaml in the current directory

kfctl build would not create an app directory but instead treat the current directory as the app dir.

This eliminates the need to cd into the app directory after running `kfctl init`. Right now the directory creation that happens in init is akward because the next two commands generate/apply then execute from the created directory.

One time `init` actions (like API enablement). Should happen during apply. 

We can do that by changing the semantics of kfctl to act like a K8s controller and treating the apply method as the reconcile method.

* kfctl will repeatedly call apply until some conditions in KfDef.Status is reached
* `app.yaml` will be updated so conditions are persisted on error
* KfDef will be passed back and forth between the KfApps so that KfApps can modify KfDef.Status

I think this would also allow post apply platform specific logic like @yanniszark suggested e.g. printing out the endpoint to access Kubeflow.

**Comment by kkasravi** [comment]
right now after `kfctl generate` you get a collection of microservices
```
admission-webhook.yaml   cert-manager.yaml        katib-ui.yaml            pytorch-job-crds.yaml
ambassador.yaml          cloud-endpoints.yaml     metacontroller.yaml      pytorch-operator.yaml
api-service.yaml         gpu-driver.yaml          metrics-collector.yaml   scheduledworkflow.yaml
application-crds.yaml    istio-crds.yaml          minio.yaml               spartakus.yaml
application.yaml         istio-install.yaml       mysql.yaml               suggestion.yaml
argo.yaml                istio.yaml               notebook-controller.yaml tensorboard.yaml
basic-auth-ingress.yaml  jupyter-web-app.yaml     persistent-agent.yaml    tf-job-operator.yaml
basic-auth.yaml          katib-controller.yaml    pipelines-runner.yaml
centraldashboard.yaml    katib-db.yaml            pipelines-ui.yaml
cert-manager-crds.yaml   katib-manager.yaml       pipelines-viewer.yaml
```
if we did something like kfctl build && kubectl apply <output-dir> then would we have 
kfapp-controllers (eventually application-controllers) deployed as well as KfApp CRs where the kfapp-controllers would claim ownership of KfApps and update their status?

**Comment by jlewi** [comment]
I guess we might do that but I think my proposal is a little simpler 

The current UX is

```
kfctl init ...
kfctl generate ...
kfctl apply ...
```
I'm just proposing we consolidate this into 

```
kfctl build
kfctl apply
```

Here's how the behavior changes

init is doing the following

  1. Doing one time configuration e.g. enabling GCP services
  1. Creating a directory 
  1. Downloading the config to app.yaml

This would be changed as follows
 
* kfctl no longer creates a directory; it just operates in the current directory
* kfctl build would download the config to the current directory
* One time configuration e.g. enabling GCP services would happen during apply
  * kfctl would use KfDef.Status.Conditions to keep track of whether one time steps had been performed
    so that they could be skipped on successive invocations of `kfctl generate`

The functionality invoked during `kfctl generate` e.g. generating YAML files given app.yaml would happen during `kfctl build`.

The primary reason to rename generate to build is to match kustomize.

**Comment by kkasravi** [comment]
ok, got it. Yes much simpler.

**Comment by jlewi** [comment]
I took another stab at clarifying what I think the semantics should look like in the design doc.
https://docs.google.com/document/d/1kviNOdjoBPeoJ56KBvMKWq7SzrSN5rFUQoULK8muD-M/edit#bookmark=id.iwxbmwl36xkz

@yanniszark @swiftdiaries @kkasravi Can we try to come to some agreement about the semantics and begin implementation? We are are rapidly running out of time for 0.7.

**Comment by jlewi** [comment]
Thanks @swiftdiaries. 

@gabrielwen do you think you will be able to pick this up?

**Comment by yanniszark** [comment]
I believe @swiftdiaries has made a PR here: https://github.com/kubeflow/kubeflow/pull/4040

**Comment by swiftdiaries** [comment]
That PR is for the apply semantics alone right now.
I can do a separate PR or bundle in the build changes in the same PR. Both work for me :)

**Comment by jlewi** [comment]
@gabrielwen PTAL

**Comment by jlewi** [comment]
Bump to P0 because this must land in 0.7.

**Comment by jlewi** [comment]
@swiftdiaries Could you provide an update on this please? Do we have an ETA for when your PR #4040 would be merged. After #4040 what would the remaining work be?

**Comment by swiftdiaries** [comment]
I'll finish up #4040 in 2 days.
Currently, #4040 is a thin wrapper around the existing code with a a little bit of extra code for the `init` part alone. It still uses the platform ResourceEnums.
The next steps would be to remove them all together and move away from them in favor of the `platform` field and plugins.

**Comment by swiftdiaries** [comment]
@jlewi @yanniszark @gabrielwen 
Can we list remaining tasks to be done before 0.7?
I'd like to start working on them.

Remaining tasks:

- [ ] coordinator cleanup
- [ ] unit tests for coordinator functions

**Comment by jlewi** [comment]
@swiftdiaries What's the difference between the last two tasks?

Other tasks
* Update the docs
* Did we already add a "build" command to replace generate?
* Should we add an E2E test that verifies that invoking build and then apply works as expected?

**Comment by swiftdiaries** [comment]
For the difference,
Even if we use the same directory, currently a lot of code references a `app.yaml` so it'd be `/some/path/kfdef.yaml` writing to `/some/path/app.yaml`
So I wanted to split that into two separate tasks. 
First would be to change the working directory for existing filesystem paths vs URLs from the internet. 
Second would be to replace the app.yaml reference with the same kfdef file.
For the build as a replacement for generate, that was already added in #4040. 
There's a bug in that implementation, so I've added a fix and added an E2E test to verify the build -> apply workflow here: https://github.com/kubeflow/kubeflow/pull/4232

**Comment by jlewi** [comment]
Thanks @swiftdiaries .

* What about the old commands "init" and "generate" have we either removed them or do we print out errors telling people to use the new commands?

**Comment by jlewi** [comment]
@swiftdiaries Any update on this?

**Comment by sarahmaddox** [comment]
@swiftdiaries A few questions from a docs point of view:

**Question 1:** Please would you describe the new sequence of commands needed to install Kubeflow?

I'm looking to update the following docs (issue https://github.com/kubeflow/website/issues/1226):

* https://www.kubeflow.org/docs/started/k8s/kfctl-k8s-istio/
```
export KFAPP="<your-choice-of-application-directory-name>"
export CONFIG="https://raw.githubusercontent.com/kubeflow/kubeflow/v0.6-branch/bootstrap/config/kfctl_k8s_istio.0.6.2.yaml"
kfctl init ${KFAPP} --config=${CONFIG} -V

cd ${KFAPP}
kfctl generate all -V
kfctl apply all -V
```

* https://www.kubeflow.org/docs/gke/deploy/deploy-cli/
```
export KFAPP=<your choice of application directory name>
export ZONE=<your target GCP zone> # where the deployment will be created
export PROJECT=<your GCP project ID>
CONFIG="https://raw.githubusercontent.com/kubeflow/kubeflow/cfb336b7/bootstrap/config/kfctl_gcp_iap.0.6.2.yaml"
kfctl init ${KFAPP} --project=${PROJECT} --config=${CONFIG} -V

cd ${KFAPP}
kfctl generate all -V --zone ${ZONE}
kfctl apply all -V
```

* Other similar docs.

I've seen the v0.7 demo script (http://bit.ly/demo-v0-7) but the info there doesn't seem to match the planned kfctl semantics, which I understand to be replacing the `kfctl init` and `kfctl generate` commands with `kfctl build`.

**Question 2:** Are we removing `kfctl init` and `kfctl generate` entirely or leaving them in place but deprecated?

**Question 3:** How much of this change will be in Kubeflow v0.7?

Many thanks
Sarah

**Comment by jlewi** [comment]
We are removing init entirely. Generate is being renamed build.

On GCP the command would be

```
export KFAPP=<your choice of Kubeflow deployment name>
export CONFIG=<http path of kfctl_gcp_iap.yaml>
export CLIENT_ID = <client id for IAP>
export CLIENT_SECRET= <client secret for IAP>

mkdir ${KFAPP}
cd ${KFAPP}
kfctl apply -V -f ${CONFIG}
```

   * Project, zone, and email are inferred automatically from gcloud settings
   * User can also explicitly set these by downloading kfctl_gcp_iap.yaml, editing it, and then calling apply on it.


Advanced mode would be

```
mkdir ${KFAPP}
cd ${KFAPP}
kfctl build -V -f ${CONFIG}
<edit the config files>
kfctl apply -V -f ${CONFIG}
```
  
   * So compared to v0.6; a user doesn't have to always run build before calling apply
   * They only have to run build if they want to generate the configs, edit them, and then apply them
   * build replaces generate

**Comment by swiftdiaries** [comment]
For the default installation on k8s,
```
export KFAPP=<kubeflow deployment name>
export CONFIG=https://raw.githubusercontent.com/kubeflow/kubeflow/master/bootstrap/config/kfctl_k8s_istio.yaml
mkdir ${KFAPP}
cd ${KFAPP}
kfctl apply -V -f ${CONFIG}
```

**Comment by swiftdiaries** [comment]
@jlewi 
Added an E2E test for both the use cases.
There are a couple of PRs out for fixes:
- [ ] https://github.com/kubeflow/kubeflow/pull/4287
- [ ] https://github.com/kubeflow/kubeflow/pull/4279
- [x] https://github.com/kubeflow/kubeflow/pull/4275
AFAIK, that should cover it for 0.7.
There is some cleanup that should be done but that'll be part of kfctl 1.0.

**Comment by sarahmaddox** [comment]
Doc updates in progress: https://github.com/kubeflow/website/pull/1234 (not yet ready for review).

**Comment by jlewi** [comment]
It looks like the generate and init commands are still defined
https://github.com/kubeflow/kubeflow/blob/master/bootstrap/cmd/kfctl/cmd/generate.go
https://github.com/kubeflow/kubeflow/blob/master/bootstrap/cmd/kfctl/cmd/init.go

Can we change these commands so that we just print out an error and exit.

The error should be something like 

The init command has been removed in 0.7. You can just run kfctl apply -f ${CONFIG} file. For more information run `kfctl apply --help or refer to the docs www.kubeflow.org` 

We should provide a similar message for generate but use the build command instead.

It also looks like the kfctl apply message needs to be fixed

```
kfctl apply --help
Deploy a generated kubeflow application.

Usage:
  kfctl apply [all(=default)|k8s|platform] [flags]

...
```

We should remove `[all(=default)|k8s|platform]`  because that is no longer a valid argument.

@swiftdiaries @lluunn @gabrielwen Could one of you take this on please?

**Comment by swiftdiaries** [comment]
PR: https://github.com/kubeflow/kubeflow/pull/4351

**Comment by jlewi** [comment]
Thanks @swiftdiaries lets verify this in the next RC and then we can close this issue.

**Comment by jlewi** [comment]
Using kubeflow/kubeflow@0ba828a18ee97094a5aa24da2fe09ca75d05d15d


```
./bootstrap/bin/kfctl apply --help
'kfctl apply' builds and deploys a kubeflow application from a KFDef config.
To install run -> kfctl apply -f ${CONFIG}
For more information, run 'kfctl apply -h' or read the docs at www.kubeflow.org.

Usage:
  kfctl apply -f ${CONFIG} [flags]

Flags:
  -f, --file string   Static config file to use. Can be either a local path or a URL.
                      For example:
                      export CONFIG=https://raw.githubusercontent.com/kubeflow/manifests/v0.7-branch/kfdef/kfctl_existing_arrikto.yaml
                      export CONFIG=https://raw.githubusercontent.com/kubeflow/manifests/v0.7-branch/kfdef/kfctl_aws.yaml
                      export CONFIG=https://raw.githubusercontent.com/kubeflow/manifests/v0.7-branch/kfdef/kfctl_gcp_iap.yaml
                      export CONFIG=https://raw.githubusercontent.com/kubeflow/manifests/v0.7-branch/kfdef/kfctl_k8s_istio.yaml
                      kfctl apply -V --file=${CONFIG}
  -h, --help          help for apply
  -V, --verbose       verbose output default is false
```

```
 ./bootstrap/bin/kfctl init --help
'kfctl init' has been removed.
Please switch to new semantics.
To install run -> kfctl apply -f ${CONFIG}
For more information, run 'kfctl apply -h' or read the docs at www.kubeflow.org.

Usage:
  kfctl init [flags]

Flags:
  -h, --help   help for init
```

```
./bootstrap/bin/kfctl generate --help
'kfctl generate' has been replaced by 'kfctl build'
Please switch to new semantics.
To build a KFAPP run -> kfctl build -f ${CONFIG}
Then to install -> kfctl apply
For more information, run 'kfctl build -h' or read the docs at www.kubeflow.org.

Usage:
  kfctl generate [flags]

Flags:
  -h, --help   help for generate
```

@swiftdiaries Everything looks. One minor thing is we have renamed the KFDef files on the v0.7.0 branch to have 0.7.0 suffix.

see
https://github.com/kubeflow/manifests/tree/v0.7-branch/kfdef

Could we update the name of the configs? Or maybe we should just include the link to
https://github.com/kubeflow/manifests/tree/v0.7-branch/kfdef and tell people to choose one of the
files there?

This way the help message won't become outdated if we do a minor release of the configs?

**Comment by jlewi** [comment]
Using: kubeflow/kubeflow - v0.4.0-rc.1-802-gd66f8b97
which corresponds to RC0.7
https://github.com/kubeflow/kubeflow/releases/tag/v0.7.0-rc.7

```
~/git_kubeflow-kubeflow/bootstrap/bin/kfctl apply --help
'kfctl apply' builds and deploys a kubeflow application from a KFDef config.
To install run -> kfctl apply -f ${CONFIG}
For more information, run 'kfctl apply -h' or read the docs at www.kubeflow.org.

Usage:
  kfctl apply -f ${CONFIG} [flags]

Flags:
  -f, --file string   Static config file to use. Can be either a local path:
                      		export CONFIG=./kfctl_gcp_iap.yaml
                      	or a URL:
                      		export CONFIG=https://raw.githubusercontent.com/kubeflow/manifests/v0.7-branch/kfdef/kfctl_gcp_iap.0.7.0.yaml
                      		export CONFIG=https://raw.githubusercontent.com/kubeflow/manifests/v0.7-branch/kfdef/kfctl_existing_arrikto.0.7.0.yaml
                      		export CONFIG=https://raw.githubusercontent.com/kubeflow/manifests/v0.7-branch/kfdef/kfctl_aws.0.7.0.yaml
                      		export CONFIG=https://raw.githubusercontent.com/kubeflow/manifests/v0.7-branch/kfdef/kfctl_k8s_istio.0.7.0.yaml
                      	kfctl apply -V --file=${CONFIG}
  -h, --help          help for apply
  -V, --verbose       verbose output default is false
```

So everything looks good. 

Marking this as fixe.d
